File: dotnet/src/Sodium/CSharp/Sodium.Frp.Tests/CellTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using NUnit.Framework;
using Sodium.Functional;

namespace Sodium.Frp.Tests
{
    [TestFixture]
    public class CellTests
    {
        [Test]
        public void TestLoop()
        {
            (Cell<int> c, CellStreamSink<int> s) = Transaction.Run(() =>
             {
                 CellLoop<int> loop = Cell.CreateLoop<int>();
                 Cell<int> cLocal = loop.Map(v => v * 5);
                 CellStreamSink<int> sLocal = Cell.CreateStreamSink<int>();
                 loop.Loop(sLocal.Hold(3));
                 return (cLocal, sLocal);
             });

            List<int> output1 = new List<int>();
            List<int> output2 = new List<int>();
            IListener l = c.Listen(output1.Add);
            IListener l2 = c.Updates().Listen(output2.Add);

            s.Send(5);
            s.Send(7);

            l2.Unlisten();
            l.Unlisten();

            CollectionAssert.AreEqual(new[] { 15, 25, 35 }, output1);
            CollectionAssert.AreEqual(new[] { 25, 35 }, output2);
        }

        [Test]
        public void TestLiftSimultaneousUpdates()
        {
            List<int> @out = new List<int>();
            CellSink<int> cellSink = Cell.CreateSink(1);
            Cell<int> cell = cellSink.Map(v => 2 * v);
            IListener l = cellSink.Lift(cell, (x, y) => x + y).Updates().Listen(@out.Add);

            cellSink.Send(2);
            cellSink.Send(7);

            l.Unlisten();

            CollectionAssert.AreEqual(new[] { 6, 21 }, @out);
        }

        [Test]
        public void TestLiftInSwitchC()
        {
            IReadOnlyList<Test> list1 = new[] { new Test(0), new Test(1), new Test(2), new Test(3), new Test(4) };
            IReadOnlyList<Test> list2 = new[] { new Test(5), new Test(6), new Test(7), new Test(8), new Test(9) };

            CellSink<IReadOnlyList<Test>> v = Cell.CreateSink(list1);

            Cell<IReadOnlyList<int>> c = v.Map(oo => oo.Select(o => o.Value).Lift()).SwitchC();

            List<IReadOnlyList<int>> streamOutput = new List<IReadOnlyList<int>>();
            IListener l = c.Updates().Listen(streamOutput.Add);

            List<IReadOnlyList<int>> cellOutput = new List<IReadOnlyList<int>>();
            IListener l2 = c.Listen(cellOutput.Add);

            list1[2].Value.Send(12);
            list2[1].Value.Send(16);
            list1[4].Value.Send(14);
            Transaction.RunVoid(() =>
            {
                list2[2].Value.Send(17);
                list1[0].Value.Send(10);
                v.Send(list2);
            });
            list1[3].Value.Send(13);
            list2[3].Value.Send(18);

            l2.Unlisten();
            l.Unlisten();

            Assert.AreEqual(4, streamOutput.Count);
            Assert.AreEqual(5, cellOutput.Count);

            CollectionAssert.AreEqual(new[] { 0, 1, 2, 3, 4 }, cellOutput[0]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 4 }, streamOutput[0]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 4 }, cellOutput[1]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 14 }, streamOutput[1]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 14 }, cellOutput[2]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 8, 9 }, streamOutput[2]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 8, 9 }, cellOutput[3]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 18, 9 }, streamOutput[3]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 18, 9 }, cellOutput[4]);
        }

        [Test]
        public void TestMapWithSwitchC()
        {
            IReadOnlyList<Test> list1 = new[] { new Test(0), new Test(1), new Test(2), new Test(3), new Test(4) };
            IReadOnlyList<Test> list2 = new[] { new Test(5), new Test(6), new Test(7), new Test(8), new Test(9) };

            CellSink<IReadOnlyList<Test>> v = Cell.CreateSink(list1);

            Cell<IReadOnlyList<int>> c = v.Map(oo => oo.Select(o => o.Value).Lift()).Map(o => o).SwitchC();

            List<IReadOnlyList<int>> streamOutput = new List<IReadOnlyList<int>>();
            IListener l = c.Updates().Listen(streamOutput.Add);

            List<IReadOnlyList<int>> cellOutput = new List<IReadOnlyList<int>>();
            IListener l2 = c.Listen(cellOutput.Add);

            list1[2].Value.Send(12);
            list2[1].Value.Send(16);
            list1[4].Value.Send(14);
            Transaction.RunVoid(() =>
            {
                list2[2].Value.Send(17);
                list1[0].Value.Send(10);
                v.Send(list2);
            });
            list1[3].Value.Send(13);
            list2[3].Value.Send(18);

            l2.Unlisten();
            l.Unlisten();

            Assert.AreEqual(4, streamOutput.Count);
            Assert.AreEqual(5, cellOutput.Count);

            CollectionAssert.AreEqual(new[] { 0, 1, 2, 3, 4 }, cellOutput[0]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 4 }, streamOutput[0]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 4 }, cellOutput[1]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 14 }, streamOutput[1]);
            CollectionAssert.AreEqual(new[] { 0, 1, 12, 3, 14 }, cellOutput[2]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 8, 9 }, streamOutput[2]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 8, 9 }, cellOutput[3]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 18, 9 }, streamOutput[3]);
            CollectionAssert.AreEqual(new[] { 5, 16, 17, 18, 9 }, cellOutput[4]);
        }

        public class Test
        {
            public Test(int initialValue)
            {
                this.Value = Cell.CreateSink(initialValue);
            }

            public CellSink<int> Value { get; }
        }

        [Test]
        public void TestLiftCellsInSwitchC()
        {
            List<int> @out = new List<int>();
            CellSink<int> s = Cell.CreateSink(0);
            Cell<Cell<int>> c = Cell.Constant(Cell.Constant(1));
            Cell<Cell<int>> r = c.Map(c2 => c2.Lift(s, (v1, v2) => v1 + v2));
            IListener l = r.SwitchC().Listen(@out.Add);
            s.Send(2);
            s.Send(4);
            l.Unlisten();
            CollectionAssert.AreEqual(new[] { 1, 3, 5 }, @out);
        }

        [Test]
        public void TestLazyCellCreation()
        {
            List<int> @out = new List<int>();
            StreamSink<int> s = Stream.CreateSink<int>();
            Cell<Cell<int>> c = Cell.Constant(1).Map(_ => s.Hold(0));
            s.Send(1);
            IListener l = c.SwitchC().Listen(@out.Add);
            s.Send(3);
            s.Send(5);
            l.Unlisten();
            CollectionAssert.AreEqual(new[] { 1, 3, 5 }, @out);
        }

        [Test]
        public void CellValuesWithPrevious()
        {
            StreamSink<int> s = Stream.CreateSink<int>();
            Cell<int> c = s.Hold(0);
            List<(int Current, Maybe<int> Previous)> @out = new List<(int Current, Maybe<int> Previous)>();
            using (Transaction.Run(
                () =>
                {
                    Stream<(int Current, Maybe<int> Previous)> r = c.Updates()
                        .Snapshot(c, (n, o) => (Current: n, Previous: Maybe.Some(o)))
                        .OrElse(
                            Cell.ConstantLazy(c.SampleLazy()).Values().Map(v => (Current: v, Previous: Maybe<int>.None)));
                    return r.Listen(@out.Add);
                }))
            {
                s.Send(1);
                s.Send(2);
                s.Send(3);
                s.Send(4);
            }

            CollectionAssert.AreEqual(
                new[]
                {
                    (Current: 0, Previous: Maybe.None),
                    (Current: 1, Previous: Maybe.Some(0)),
                    (Current: 2, Previous: Maybe.Some(1)),
                    (Current: 3, Previous: Maybe.Some(2)),
                    (Current: 4, Previous: Maybe.Some(3))
                },
                @out);
        }

        [Test]
        public void CellValuesWithPreviousHavingInitialUpdate()
        {
            StreamSink<int> s = Stream.CreateSink<int>();
            Cell<int> c = s.Hold(0);
            List<(int Current, Maybe<int> Previous)> @out = new List<(int Current, Maybe<int> Previous)>();
            using (Transaction.Run(
                () =>
                {
                    Stream<(int Current, Maybe<int> Previous)> r = c.Updates()
                        .Snapshot(c, (n, o) => (Current: n, Previous: Maybe.Some(o)))
                        .OrElse(
                            Cell.ConstantLazy(c.SampleLazy()).Values().Map(v => (Current: v, Previous: Maybe<int>.None)));
                    s.Send(1);
                    return r.Listen(@out.Add);
                }))
            {
                s.Send(2);
                s.Send(3);
                s.Send(4);
                s.Send(5);
            }

            CollectionAssert.AreEqual(
                new[]
                {
                    (Current: 1, Previous: Maybe.Some(0)),
                    (Current: 2, Previous: Maybe.Some(1)),
                    (Current: 3, Previous: Maybe.Some(2)),
                    (Current: 4, Previous: Maybe.Some(3)),
                    (Current: 5, Previous: Maybe.Some(4))
                },
                @out);
        }

        [Test]
        public void TestLoopAndSwitchCError()
        {
            InvalidOperationException exception = null;
            try
            {
                Cell.Loop<int>()
                    // ReSharper disable once ReturnValueOfPureMethodIsNotUsed
                    .WithoutCaptures(
                        l =>
                        {
                            StreamSink<Inner> s = Stream.CreateSink<Inner>();
                            Cell<Inner> cc = s.Hold(new Inner(l.SampleLazy()));
                            Cell<int> c = cc.Map(o => o.C).SwitchC();
                            return c;
                        });
            }
            catch (InvalidOperationException e)
            {
                exception = e;
            }
            
            Assert.IsNotNull(exception);
            Assert.AreEqual("ValueFactory attempted to access the Value property of this instance.", exception.Message);
        }

        [Test]
        public void TestLoopAndSwitchC()
        {
            (Cell<int> resultCell, (Cell<Inner> innerCell, StreamSink<Inner> innerStreamSink)) =
                Cell.Loop<int>()
                    .WithCaptures(
                        l =>
                        {
                            StreamSink<Inner> s = Stream.CreateSink<Inner>();
                            Cell<Inner> cc = s.Hold(new Inner(l.SampleLazy()));
                            Cell<int> c = cc.Map(o => o.C).SwitchC().Values().Hold(3);
                            return (Cell: c, Captures: (cc, s));
                        });

            List<int> @out = new List<int>();
            using (resultCell.Listen(@out.Add))
            {
                innerCell.Sample().S.Send(5);
                innerStreamSink.Send(new Inner(resultCell.SampleLazy().Map(v => v - 1)));
                innerCell.Sample().S.Send(7);
            }

            CollectionAssert.AreEqual(new[] { 3, 5, 4, 7 }, @out);
        }

        private class Inner
        {
            public Inner(Lazy<int> initialValue)
            {
                this.S = Stream.CreateSink<int>();
                this.C = this.S.HoldLazy(initialValue);
            }

            public StreamSink<int> S { get; }
            public Cell<int> C { get; }
        }
    }
}

File: dotnet/src/Sodium/FSharp/Sodium.FSharp.Frp.Tests/CellTests.fs
module Sodium.Frp.Tests.Cell

open System
open System.Collections.Generic
open NUnit.Framework
open Sodium.Frp

[<AutoOpen>]
module private Types =
    type Test (initialValue : int) = member val Value = sinkC initialValue
    type Inner = private { s' : StreamSink<int>; c' : Cell<int> }
        with
            static member create initialValue =
                let s = sinkS ()
                { s' = s; c' = s |> holdLazyS initialValue }
            member this.s = this.s'
            member this.c = this.c'

[<TestFixture>]
type ``Cell Tests``() =

    [<Test>]
    member __.``Test Loop``() =
        let struct (_, (c, s)) = loopC (fun cl ->
            let c = cl |> mapC ((*) 5)
            let s = sinkCS<int> ()
            struct (s |> holdS 3, (c, s)))
        let output1 = List<_>()
        let output2 = List<_>()
        let l = c |> listenC output1.Add
        let l2 = c |> updatesC |> listenS output2.Add
        s |> sendS 5
        s |> sendS 7
        l2 |> unlistenL
        l |> unlistenL
        CollectionAssert.AreEqual([15;25;35], output1)
        CollectionAssert.AreEqual([25;35], output2)

    [<Test>]
    member __.``Test Lift Simultaneous Updates``() =
        let out = List<_>()
        let cellSink = sinkC 1
        let cell = cellSink |> mapC ((*) 2)
        let l = (cellSink, cell) |> lift2C (+) |> updatesC |> listenS out.Add
        cellSink |> sendC 2
        cellSink |> sendC 7
        l |> unlistenL
        CollectionAssert.AreEqual([6;21],out)

    [<Test>]
    member __.``Test Lift In SwitchC``() =
        let list1 = [|Test(0);Test(1);Test(2);Test(3);Test(4)|]
        let list2 = [|Test(5);Test(6);Test(7);Test(8);Test(9)|]
        let v = sinkC list1
        let c = v |> mapC (Seq.map (fun o -> o.Value) >> liftAllC id) |> switchC
        let streamOutput = List<_>()
        let l = c |> updatesC |> listenS streamOutput.Add
        let cellOutput = List<_>()
        let l2 = c |> listenC cellOutput.Add
        list1.[2].Value |> sendC 12
        list2.[1].Value |> sendC 16
        list1.[4].Value |> sendC 14
        runT (fun () ->
            list2.[2].Value |> sendC 17
            list1.[0].Value |> sendC 10
            v |> sendC list2)
        list1.[3].Value |> sendC 13
        list2.[3].Value |> sendC 18
        l2 |> unlistenL
        l |> unlistenL
        Assert.AreEqual (4, streamOutput.Count)
        Assert.AreEqual (5, cellOutput.Count)
        CollectionAssert.AreEqual ([0;1;2;3;4], cellOutput.[0])
        CollectionAssert.AreEqual ([0;1;12;3;4], streamOutput.[0])
        CollectionAssert.AreEqual ([0;1;12;3;4], cellOutput.[1])
        CollectionAssert.AreEqual ([0;1;12;3;14], streamOutput.[1])
        CollectionAssert.AreEqual ([0;1;12;3;14], cellOutput.[2])
        CollectionAssert.AreEqual ([5;16;17;8;9], streamOutput.[2])
        CollectionAssert.AreEqual ([5;16;17;8;9], cellOutput.[3])
        CollectionAssert.AreEqual ([5;16;17;18;9], streamOutput.[3])
        CollectionAssert.AreEqual ([5;16;17;18;9], cellOutput.[4])

    [<Test>]
    member __.``Test Map With SwitchC``() =
        let list1 = [|Test(0);Test(1);Test(2);Test(3);Test(4)|]
        let list2 = [|Test(5);Test(6);Test(7);Test(8);Test(9)|]
        let v = sinkC list1
        let c = v |> mapC ((Seq.map (fun o -> o.Value)) >> liftAllC id) |> mapC id |> switchC
        let streamOutput = List<_>()
        let l = c |> updatesC |> listenS streamOutput.Add
        let cellOutput = List<_>()
        let l2 = c |> listenC cellOutput.Add
        list1.[2].Value |> sendC 12
        list2.[1].Value |> sendC 16
        list1.[4].Value |> sendC 14
        runT (fun () ->
            list2.[2].Value |> sendC 17
            list1.[0].Value |> sendC 10
            v |> sendC list2)
        list1.[3].Value |> sendC 13
        list2.[3].Value |> sendC 18
        l2 |> unlistenL
        l |> unlistenL
        Assert.AreEqual (4, streamOutput.Count)
        Assert.AreEqual (5, cellOutput.Count)
        CollectionAssert.AreEqual ([0;1;2;3;4], cellOutput.[0])
        CollectionAssert.AreEqual ([0;1;12;3;4], streamOutput.[0])
        CollectionAssert.AreEqual ([0;1;12;3;4], cellOutput.[1])
        CollectionAssert.AreEqual ([0;1;12;3;14], streamOutput.[1])
        CollectionAssert.AreEqual ([0;1;12;3;14], cellOutput.[2])
        CollectionAssert.AreEqual ([5;16;17;8;9], streamOutput.[2])
        CollectionAssert.AreEqual ([5;16;17;8;9], cellOutput.[3])
        CollectionAssert.AreEqual ([5;16;17;18;9], streamOutput.[3])
        CollectionAssert.AreEqual ([5;16;17;18;9], cellOutput.[4])

    [<Test>]
    member __.``Test Lift Cells In SwitchC``() =
        let out = List<_>()
        let s = sinkC 0
        let c = constantC <| constantC 1
        let r = c |> mapC (fun c -> (c, s) |> lift2C (+))
        let l = r |> switchC |> listenC out.Add
        s |> sendC 2
        s |> sendC 4
        l |> unlistenL

    [<Test>]
    member __.``Test Lazy Cell Creation``() =
        let out = List<_>()
        let s = sinkS ()
        let c = constantC 1 |> mapC (fun _ -> s |> holdS 0)
        s |> sendS 1
        let l = c |> switchC |> listenC out.Add
        s |> sendS 3
        s |> sendS 5
        l |> unlistenL
        CollectionAssert.AreEqual([1;3;5],out)

    [<Test>]
    member __.``Test Cell Values With Previous``() =
        let s = sinkS ()
        let c = s |> holdS 0
        let out = List<_>()
        let l = runT (fun () ->
            let r = (c |> updatesC |> snapshotC c (fun c p -> (c, Some p)), constantLazyC (c |> sampleLazyC) |> valuesC |> mapS (fun c -> (c, None))) |> orElseS
            r |> listenS out.Add)
        s |> sendS 1
        s |> sendS 2
        s |> sendS 3
        s |> sendS 4
        l |> unlistenL
        CollectionAssert.AreEqual([(0,None);(1,Some 0);(2,Some 1);(3,Some 2);(4,Some 3)],out)

    [<Test>]
    member __.``Test Cell Values With Previous Having Initial Update``() =
        let s = sinkS ()
        let c = s |> holdS 0
        let out = List<_>()
        let l = runT (fun () ->
            let r = (c |> updatesC |> snapshotC c (fun c p -> (c, Some p)), constantLazyC (c |> sampleLazyC) |> valuesC |> mapS (fun c -> (c, None))) |> orElseS
            s |> sendS 1
            r |> listenS out.Add)
        s |> sendS 2
        s |> sendS 3
        s |> sendS 4
        s |> sendS 5
        l |> unlistenL
        CollectionAssert.AreEqual([(1,Some 0);(2,Some 1);(3,Some 2);(4,Some 3);(5,Some 4)],out)

    [<Test>]
    member __.``Test Loop And SwitchC Error``() =
        let e =
            try
                loopWithNoCapturesC (fun c ->
                    let s = sinkS ()
                    let cc = s |> holdS (Inner.create (c |> sampleLazyC))
                    cc |> mapC (fun o -> o.c) |> switchC) |> ignore
                None
            with
                | :? InvalidOperationException as e -> Some e
        e |> assertExceptionExists (fun e -> Assert.AreEqual ("ValueFactory attempted to access the Value property of this instance.", e.Message))

    [<Test>]
    member __.``Test Loop And SwitchC``() =
        let struct (resultCell, (innerCell, innerStreamSink)) =
            loopC (fun c ->
                let s = sinkS ()
                let cc = s |> holdS (Inner.create (c |> sampleLazyC))
                struct (cc |> mapC (fun o -> o.c) |> switchC |> valuesC |> holdS 3, (cc, s)))
        let out = List<_>()
        (
            use _l = resultCell |> listenC out.Add
            (innerCell |> sampleC).s |> sendS 5
            innerStreamSink |> sendS (Inner.create (resultCell |> sampleLazyC |> Lazy.map (fun v -> v - 1)))
            (innerCell |> sampleC).s |> sendS 7
        )
        CollectionAssert.AreEqual([3;5;4;7],out)